### PR TITLE
[deckhouse-config] fix: apply defaults before spec.settings validation

### PR DESCRIPTION
**Pull request by diafour**
## Description

Validate raw spec.settings is not compatible with addon-operator behaviour. Addon-operator applies defaults before validating "config values" (see [module_manager.go](https://github.com/flant/addon-operator/blob/main/pkg/module_manager/module_manager.go#L979-L981)).

## Why do we need it, and what problem does it solve?

deckhouse-config module treats valid spec.settings as invalid and disables previously enabled module.

## What is the expected result?

Should not disable modules on update v1.40 -> v1.41.

## Checklist
- [ ] The code is covered by unit tests.
- [ ] e2e tests passed.
- [ ] Documentation updated according to the changes.
- [ ] Changes were tested in the Kubernetes cluster manually.

## Changelog entries

```changes
section: deckhouse-config
type: fix
summary: Apply defaults before spec.settings validation.
impact_level: default
```

### CODE REVIEW

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @diafour. I've started the workflow for label 'e2e/run/aws'...

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @diafour. I've started the workflow for label 'e2e/run/aws'...

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @diafour. I've started the workflow for label 'e2e/run/azure'...

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @diafour. I've started the workflow for label 'e2e/run/yandex-cloud'...

File: go_lib/deckhouse-config/validator.go [hunk 1]
```diff
@@ -168,18 +172,32 @@ func (c *ConfigValidator) Validate(cfg *d8cfg_v1alpha1.ModuleConfig) ValidationR
...
+	}
+
+	// Merge defaults with passed settings as addon-operator will do.
+	values := utils.MergeValues(
+		utils.Values{valuesKey: defaultSettings},
+		utils.Values{valuesKey: cfgSettings},
+	)
+
 	if cfgName == "global" {
 		return c.valuesValidator.ValidateGlobalConfigValues(values)
```
Comment (shvgn) [inline go_lib/deckhouse-config/validator.go:202]: Why do we treat global values and module values separately? It's fine that we load them from different places. But it looks like unnecessary interface bloat to me.

It's more for future cosideration than for addressing in current PR.

**Review by shvgn — DISMISSED** [review]

File: go_lib/deckhouse-config/validator.go (hunk 1 above)
Comment (diafour) [reply go_lib/deckhouse-config/validator.go:202]: Yeah, it should be addressed in the Addon-operator. I've choose no to wrap these methods and use them as-is. See [validator.go](https://github.com/flant/addon-operator/blob/main/pkg/values/validation/validator.go#L25-L39)

File: go_lib/deckhouse-config/initial_config_loader.go [hunk 2]
```diff
@@ -52,22 +52,28 @@ func NewInitialConfigLoader(kubeClient client.Client) *InitialConfigLoader {
...
+	// Init Kubernetes client if it was not specified.
 	err := l.initKubeClient()
-	log.SetLevel(lvl)
 	if err != nil {
 		return nil, fmt.Errorf("init default Kubernetes client: %v", err)
 	}
 
-	// Prepare possible names from the ConfigMap.
-	// Also, return nil if the ConfigMap is not exists or contains no settings — it'll be handled by the global hook.
+	// Get ConfigMap. Return nil if the ConfigMap is not exists or contains no settings — it'll be handled by the global hook.
```
[the remark this answers is not in the document]
Comment (diafour) [reply go_lib/deckhouse-config/initial_config_loader.go:76]: Fix comment

File: global-hooks/deckhouse-config/startup_sync.go [hunk 3]
```diff
@@ -221,25 +221,29 @@ func syncModuleConfigs(input *go_hook.HookInput, generatedCM *v1.ConfigMap, allC
...
 	if generatedCM == nil || len(generatedCM.Data) == 0 {
 		return nil
 	}
 
-	// Log deleted sections.
+	// Log deleted sections in source CM.
 	for name := range generatedCM.Data {
 		if _, has := cmData[name]; !has {
-			input.LogEntry.Warnf("ModuleConfig/%s was deleted. Section '%s' will be deleted from cm/%s.", name, name, cm.Name)
+			input.LogEntry.Warnf("ModuleConfig/%s was deleted. Section '%s' will be deleted from cm/%s.", name, name, regeneratedCM.Name)
```
[the remark this answers is not in the document]
Comment (diafour) [reply global-hooks/deckhouse-config/startup_sync.go:246]: 1. Actually spec.settings and spec.version may be cleared in the ModuleConfig. 

I've rewrite this code. It is now only one message with all deleted fields.

**Review by shvgn — APPROVED** [review]

**Comment by deckhouse-BOaTswain** [comment]
Cherry pick PR [3240](https://github.com/deckhouse/deckhouse/pull/3240) to the branch [release-1.41](https://github.com/deckhouse/deckhouse/tree/release-1.41) successful!

**Pull request merged**